lib/streamlit/type_util.py

```diff
@@ -428,3 +430,31 @@ def is_version_less_than(v1: str, v2: str) -> bool:
     from packaging import version
 
     return version.parse(v1) < version.parse(v2)
+
+
+def async_generator_to_sync(
+    async_gen: AsyncGenerator[Any, Any],
+) -> Generator[Any, Any, Any]:
+    """Convert an async generator to a synchronous generator."""
+    import asyncio
+
+    close_loop = False
+    try:
+        loop = asyncio.get_running_loop()
```

Comment:
I think it might make sense to try adding a test where the async loop exists (which is the case e.g. in SiS)

Reply:
I played around a bit more with it and I think the usage of the existing event loop isn't something that is actually used in Streamlit. The event loop in SiS / Streamlit is not running in the user thread, so we won't be able to reuse this one. There might be situations where the user code already has an event loop, but we can wait for this scenario to come up in our issues and better understand the scenario.

Reply:
I simplified the method for now to always create a local event loop and tested it in SiS as well.

Reply:
Great! Did you test it by copying the code into the app code itself?

Reply:
yep, just copied the `async_generator_to_sync` implementation into an app in SiS and tried it out with `write_stream`.